apps/src/lib/node/ledger/shell/vote_extensions/bridge_pool_vext.rs

```diff
@@ -0,0 +1,602 @@
+        >,
+    > + 'iter {
+        vote_extensions.into_iter().map(|vote_extension| {
+            self.validate_bp_roots_vext_and_get_it_back(
+                vote_extension,
+                self.storage.get_current_decision_height(),
+            )
+        })
+    }
+
+    /// Takes a list of signed Bridge pool root vote extensions,
+    /// and filters out invalid instances. This also de-dduplicates
+    /// the iterator to be unique per validator address.
+    #[inline]
+    pub fn filter_invalid_bp_roots_vexts<'iter>(
+        &'iter self,
+        vote_extensions: impl IntoIterator<Item = Signed<bridge_pool_roots::Vext>>
+        + 'iter,
+    ) -> impl Iterator<Item = (token::Amount, Signed<bridge_pool_roots::Vext>)> + 'iter
+    {
+        self.validate_bp_roots_vext_list(vote_extensions)
+            .filter_map(|ext| ext.ok())
+            .dedup_by(|(_, ext_1), (_, ext_2)| {
+                ext_1.data.validator_addr == ext_2.data.validator_addr
+            })
+    }
+
+    /// Compresses a set of signed Bridge pool roots into a single
+    /// [`bridge_pool_roots::MultiSignedVext`], whilst filtering invalid
+    /// [`Signed<bridge_pool_roots::Vext>`] instances in the process.
+    #[cfg(feature = "abcipp")]
+    pub fn compress_bridge_pool_roots(
+        &self,
+        vote_extensions: Vec<Signed<bridge_pool_roots::Vext>>,
+    ) -> Option<bridge_pool_roots::MultiSignedVext> {
+        let vexts_epoch =
+            self.storage.get_epoch(self.storage.last_height).expect(
+                "The epoch of the last block height should always be known",
+            );
+        let total_voting_power =
+            u64::from(self.storage.get_total_voting_power(Some(vexts_epoch)));
+        let mut voting_power = FractionalVotingPower::default();
+
+        let mut bp_root_sigs = bridge_pool_roots::MultiSignedVext::default();
+
+        for (validator_voting_power, vote_extension) in
+            self.filter_invalid_bp_roots_vexts(vote_extensions)
+        {
+            // update voting power
+            let validator_voting_power = u64::from(validator_voting_power);
+            voting_power += FractionalVotingPower::new(
+                validator_voting_power,
+                total_voting_power,
+            )
+            .expect(
+                "The voting power we obtain from storage should always be \
+                 valid",
+            );
+            tracing::debug!(
+                ?vote_extension.sig,
+                ?vote_extension.data.validator_addr,
+                "Inserting signature into bridge_pool_roots::MultSignedVext"
+            );
+            bp_root_sigs.insert(vote_extension);
+        }
+        if voting_power <= FractionalVotingPower::TWO_THIRDS {
+            tracing::error!(
+                "Tendermint has decided on a block including Ethereum events \
+                 reflecting <= 2/3 of the total stake"
+            );
+            return None;
+        }
+        Some(bp_root_sigs)
```

Comment:
style comment: why not 
```suggestion
            None
        } else {
            Some(bp_root_sigs)
        }
```